src/libraries/System.Private.CoreLib/src/System/Decimal.DecCalc.cs

```diff
@@ -186,35 +188,74 @@ private static void UInt64x64To128(ulong a, ulong b, ref DecCalc result)
                 result.High = (uint)high;
             }
 
+            // Do partial divide for the case where (left >> 32) < den
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private static (uint Quotient, uint Remainder) Div64By32(ulong dividend, uint den)
+            {
+                if (X86.X86Base.IsSupported)
+                {
+                    return X86.X86Base.DivRem((uint)dividend, (uint)(dividend >> 32), den);
+                }
+                else
+                {
+                    // TODO: https://github.com/dotnet/runtime/issues/5213
+                    uint quo = (uint)(dividend / den);
+                    return (quo, (uint)dividend - quo * den);
+                }
+            }
+
             /// <summary>
             /// Do full divide, yielding 96-bit result and 32-bit remainder.
             /// </summary>
             /// <param name="bufNum">96-bit dividend as array of uints, least-sig first</param>
             /// <param name="den">32-bit divisor</param>
             /// <returns>Returns remainder. Quotient overwrites dividend.</returns>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
             private static uint Div96By32(ref Buf12 bufNum, uint den)
             {
-                // TODO: https://github.com/dotnet/runtime/issues/5213
-                ulong tmp, div;
-                if (bufNum.U2 != 0)
-                {
-                    tmp = bufNum.High64;
-                    div = tmp / den;
-                    bufNum.High64 = div;
-                    tmp = ((tmp - (uint)div * den) << 32) | bufNum.U0;
+                if (X86.X86Base.IsSupported)
+                {
+                    uint remainder = 0;
+
+                    if (bufNum.U2 != 0)
+                        goto Div3Word;
+                    if (bufNum.U1 >= den)
+                        goto Div2Word;
+
+                    remainder = bufNum.U1;
+                    bufNum.U1 = 0;
+                    goto Div1Word;
+Div3Word:
+                    (bufNum.U2, remainder) = X86.X86Base.DivRem(bufNum.U2, remainder, den);
+Div2Word:
+                    (bufNum.U1, remainder) = X86.X86Base.DivRem(bufNum.U1, remainder, den);
+Div1Word:
+                    (bufNum.U0, remainder) = X86.X86Base.DivRem(bufNum.U0, remainder, den);
```

Comment:
nit: we should really use a `switch` so a jump table can be generated rather than a series of labels